src/rust/plugin-work-queue/src/psql_queue.rs

```diff
@@ -0,0 +1,305 @@
+    pub async fn get_analyzer_message(&self) -> Result<Option<Message>, PsqlQueueError> {
+        // `get_message` does a few things
+        // 1. It attempts to get a message from the queue
+        //      -> Where that message isn't over a day old
+        //      -> Where that message is "visible"
+        //      -> Where that message isn't currently being evaluated by another transaction
+        //      -> Where that message is in the 'enqueued' state
+        // 2. Updates the `try_count`
+        // 3. Updates the `visible_after`
+
+        // Note that:
+        // * messages are invisible for 10 seconds
+        // * messages 'expire' after one day
+        // * messages currently do not have a maximum retry limit
+        // * The one day expiration matches our 1 day partitioning strategy
+
+        // In the future we can leverage a maximum retry limit as well as a batch version of this query
+        // A more dynamic visibility strategy would also be reasonable
+        let request: Option<NextExecutionRequest> = sqlx::query_as!(
+            NextExecutionRequest, r#"
+            UPDATE plugin_work_queue.analyzer_plugin_executions
+            SET
+                try_count  = plugin_work_queue.analyzer_plugin_executions.try_count + 1,
+                last_updated = CURRENT_TIMESTAMP,
+                visible_after  = CURRENT_TIMESTAMP + INTERVAL '10 seconds'
+            FROM (
+                 SELECT execution_key, plugin_id, pipeline_message, status, creation_time, visible_after, tenant_id
+                 FROM plugin_work_queue.analyzer_plugin_executions
+                 WHERE status = 'enqueued'
+                   AND creation_time >= (CURRENT_TIMESTAMP - INTERVAL '1 day')
+                   AND (visible_after IS NULL OR visible_after <= CURRENT_TIMESTAMP)
+                 ORDER BY creation_time ASC
+                 FOR UPDATE SKIP LOCKED
+                 LIMIT 1
+             ) AS next_execution
+             WHERE plugin_work_queue.analyzer_plugin_executions.execution_key = next_execution.execution_key
+             RETURNING
+                 next_execution.execution_key as "execution_key!: ExecutionId",
+                 next_execution.plugin_id,
+                 next_execution.pipeline_message,
+                 next_execution.tenant_id
+        "#).fetch_optional(&self.pool)
+            .await?;
+
+        Ok(request.map(|request| Message { request }))
+    }
+
+    #[instrument(err)]
+    pub async fn ack_generator(
+        &self,
+        execution_key: ExecutionId,
+        status: Status,
+    ) -> Result<(), PsqlQueueError> {
+        sqlx::query!(
+            r#"
+                UPDATE plugin_work_queue.generator_plugin_executions
+                SET status = $2,
+                    last_updated = CASE
+                        WHEN status != 'processed'
+                            THEN CURRENT_TIMESTAMP
+                            ELSE last_updated
+                        END
```

Comment:
Would it be possible to invert the sense of this comparison to make it easier to think about? (negations are always a little trickier to parse):
```sql
WHEN status = 'processed'
THEN last_updated
ELSE CURRENT_TIMESTAMP
```
Also, how does the data flow work such that the existing status would _not_ be 'processed'? Leaving the `last_updated` field unchanged seems to imply that everything has already been handled.

Reply:
I can invert it.

As for the data flow, the status starts off as 'enqueued'. The 'ack' will change it to either 'processed' or 'failed'.

I check for '!= processed' first because we don't want to update the status for a message that raced ie: if another consumer already processed this message we want *that* 'last_updated' time.

So this basically is a way of expressing "Either update it or leave it alone", which is why I had the logic inverted - the interesting cast is when status != 'processed'.

Reply:
If one worker acquires a job and hangs for the full visibility timeout another worker can acquire that same job. At that point both workers could attempt to ack the job.

Reply:
👍 makes sense